lib/client/pool_test.go

```diff
@@ -0,0 +1,244 @@
+// Test 2: IPv4 with default values for IPIP, NATOut, IPAM - assert the returned Metadata and Spec values.
+// Test 3: IPv6 with non-default bools for IPIP, NATOut, IPAM - assert the returned Metadata and Spec values.
+
+package client
+
+import (
+	"fmt"
+	"net"
+	"os"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	"github.com/projectcalico/libcalico-go/lib/api"
+	"github.com/projectcalico/libcalico-go/lib/api/unversioned"
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+	cnet "github.com/projectcalico/libcalico-go/lib/net"
+)
+
+var _ = FDescribe("Pool tests", func() {
+
+	// Describe and call convertAPItoKVPairTest.
+	Describe("Pool API to KVPair tests", func() {
+		convertAPItoKVPairTest()
+	})
+
+	// Describe and call convertKVPairtoAPITest.
+	Describe("Pool KVPair to API tests", func() {
+		convertKVPairtoAPITest()
+	})
+})
+
+// convertKVPairtoAPITest passes different values of CIDR, IPIPInterface, NATOut (NATOutgoing)
+// and Disabled to convertKVPairToAPI function (function-under-test) and asserts the output values.
+func convertKVPairtoAPITest() {
+
+	// table describes and initializes the table for TDD (Table Driven Test) input and expected output.
+	table := []struct {
+		description   string
+		CIDR          net.IPNet
+		IPIPInterface string
+		NATOut        bool
+		Disabled      bool
+		expectedIPIP  bool
+	}{
+		{
+			description: "For IPv4 with non-default values for NATOut, Disabled and IPIPInterface",
+			CIDR: net.IPNet{
+				IP:   net.IPv4(10, 0, 0, 0),
+				Mask: net.CIDRMask(24, 32),
+			},
+			IPIPInterface: "tunl0",
+			NATOut:        true,
+			Disabled:      true,
+			expectedIPIP:  true,
+		},
+		{
+			description: "For IPv4 with default values for NATOut, Disabled and IPIPInterface",
+			CIDR: net.IPNet{
+				IP:   net.IPv4(10, 0, 0, 0),
+				Mask: net.CIDRMask(24, 32),
+			},
+			IPIPInterface: "",
+			NATOut:        false,
+			Disabled:      false,
+			expectedIPIP:  false,
+		},
+		{
+			description: "For IPv6 with non-default values for NATOut, Disabled and IPIPInterface",
+			CIDR: net.IPNet{
+				IP:   net.ParseIP("fe80::00"),
+				Mask: net.CIDRMask(120, 128),
+			},
+			IPIPInterface: "tunl0",
+			NATOut:        true,
+			Disabled:      true,
+			expectedIPIP:  true,
+		},
+	}
+
+	// Go through all the values in table and test them one at a time.
+	for _, v := range table {
+		fmt.Fprintf(GinkgoWriter, "Testing KVPairtoAPI: %s\n", v.description)
+
+		It("", func() {
+
+			// Initialize KVPair struct with test arguments.
+			kvp := model.KVPair{
+				Value: &model.Pool{
+					CIDR:          cnet.IPNet{v.CIDR},
+					IPIPInterface: v.IPIPInterface,
+					Masquerade:    v.NATOut,
+					Disabled:      v.Disabled,
+				},
+			}
+
+			// Create a new client.
+			client, _ := newClient("")
+
+			p := pools{
+				c: client,
+			}
+
+			// Call function under test with test arguments and get the result.
+			out, err := p.convertKVPairToAPI(&kvp)
+
+			// Assert output to the expected values.
```

Comment:
Can we assert the Spec and Metadata equal each other as well?  I'm worried that when we add a new field we may not remember to add these explicit checks - so having a struct comparison would avoid that ... perhaps there is a test utility that does this?